### PR TITLE
Refine errors for pretrained objects

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR breaks refines the error returned for pretrained objects using the error codes now returned by hf.co to clearly indicate to the user if:
- there is a problem in the repo name
- there is a problem in the revision
- there is a problem with a specific file
- there is a general connection error

Whenever we are in a situation where the model has a version in another framework, a helpful error message is shown.

Also note that this PR removes the hard failure on `get_list_of_files` whenever there is a connection problem, which should make the CI way less flaky.

### CODE REVIEW

File: src/transformers/file_utils.py [hunk 1]
```diff
@@ -1900,6 +1900,37 @@ def http_user_agent(user_agent: Union[Dict, str, None] = None) -> str:
...
+def _raise_for_status(request):
+    """
+    Internal version of `request.raise_for_status()` that will refine a potential HTTPError.
+    """
+    if "X-Error-Code" in request.headers:
+        error_code = request.headers["X-Error-Code"]
+        if error_code == "RepoNotFound":
+            raise RepositoryNotFoundError(f"404 Client Error: Repository Not Found for url: {request.url}")
+        elif error_code == "EntryNotFound":
+            raise EntryNotFoundError(f"404 Client Error: Entry Not Found for url: {request.url}")
```
Comment (julien-c) [inline src/transformers/file_utils.py:1927]: maybe more user-friendly if you replace in the error message Entry by something like:
> Repo exists but does not contained the file you tried to get

File: src/transformers/file_utils.py [hunk 2]
```diff
@@ -1900,6 +1900,37 @@ def http_user_agent(user_agent: Union[Dict, str, None] = None) -> str:
...
+        elif error_code == "RevisionNotFound":
+            raise RevisionNotFoundError((f"404 Client Error: Revision Not Found for url: {request.url}"))
```
Comment (julien-c) [inline src/transformers/file_utils.py:1929]: > Repo exists, but the git revision you tried to get doesn't

File: src/transformers/configuration_utils.py [hunk 3]
```diff
@@ -578,30 +583,50 @@ def get_config_dict(
...
+                f"{pretrained_model_name_or_path} does not appear to have a file named {configuration_file}."
+            )
+        except HTTPError as err:
+            logger.error(err)
+            raise EnvironmentError(
+                "We couldn't connect to 'https://huggingface.co/' to load this model and it looks like "
+                f"{pretrained_model_name_or_path} is not the path to a directory conaining a {CONFIG_NAME} file.\n"
+                "Checkout your internert connection or see how to run the library in offline mode at "
+                "'https://huggingface.co/docs/transformers/installation#offline-mode'."
+            )
```
Comment (julien-c) [inline src/transformers/configuration_utils.py:613]: haha ok i hadn't seen those yet

**Review by julien-c — COMMENTED** [review]
looks really awesome so far! cc'ing @SBrandeis who built the feature server-side (and @Pierrci)

File: src/transformers/file_utils.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/file_utils.py:1927]: I tried to stay super close to the regular HTTPError messages for those. Extended version comes later on :-)

**Review by LysandreJik — APPROVED** [review]
Yes, this looks great! Good for me to apply to the rest of the objects as well.

File: src/transformers/configuration_utils.py [hunk 4]
```diff
@@ -578,30 +581,50 @@ def get_config_dict(
...
             )
-            # Load config dict
-            config_dict = cls._dict_from_json_file(resolved_config_file)
 
+        except RepositoryNotFoundError as err:
+            logger.error(err)
+            raise EnvironmentError(
+                f"{pretrained_model_name_or_path} is not a local folder and is not a valid model identifier listed on "
+                "'https://huggingface.co/models'\nIf this is a private repository, make sure you are properly logged "
+                "in with a token having permission to this repo."
```
Comment (patrickvonplaten) [inline src/transformers/configuration_utils.py:590]:
```suggestion
                "in with a token having permission to this repo (`huggingface-cli login`).
```
(nit) maybe nice to have the login command right there

**Review by patrickvonplaten — APPROVED** [review]
Looks great!

File: src/transformers/configuration_utils.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/configuration_utils.py:590]: Actually the token needs to be passed explicitly or with `use_auth_token=True`, will amend.

**Review by LysandreJik — COMMENTED** [review]
Love the new messages which tell you what files exist on the hub :)

**Review by LysandreJik — APPROVED** [review]
Thanks for applying it to the rest! LGTM

**Pull request merged**